src/v/kafka/client/client.cc

```diff
@@ -102,8 +102,9 @@ ss::future<> client::stop() noexcept {
     co_await _gate.close();
     co_await catch_and_log([this]() { return _producer.stop(); });
     for (auto& [id, group] : _consumers) {
-        for (auto& consumer : group) {
-            co_await catch_and_log([consumer]() { return consumer->leave(); });
+        while (!group.empty()) {
+            auto c = *group.begin();
+            co_await catch_and_log([c]() { return c->leave(); });
```

Comment:
It's a bit awkward, I agree. They're here: https://github.com/vectorizedio/redpanda/pull/3584/files#diff-0689c3dc4bc91cb454e761bf4a2758ef870a25c4da7cb8b8717a8cb5e0ba8f85R366-R368
```c++
          auto on_stopped = [this, group_id](const member_id& name) {
              _consumers[group_id].erase(name);
          };
```
I spent a while trying to figure out something with `boost::intrusive`, but it got tricky and I couldn't quite figure out how to register a hook that lived inside the consumer so that it can call `unlink`, since it's a behind a `shared_ptr`.
```c++
    absl::node_hash_map<
      kafka::group_id,
      boost::intrusive::unordered_set<
        consumer_entry,
        boost::intrusive::member_hook<
          consumer_entry,
          consumer_entry::hook,
          &consumer_entry::_hook>,
        boost::intrusive::constant_time_size<false>,
        boost::intrusive::hash<entry_hash>,
        boost::intrusive::equal<entry_eq>>>
      _consumers;
```
Open to suggestions.

Reply:
I don't have any suggestions, but since you are using shared_ptr you might find weak_ptr to be helpful. but in any case for this PR it'd be great if you could make a note some where about how the removal works (maybe by the loop itself, or elsewhere).